core/src/main/java/org/jboss/pnc/build/finder/core/PncBuildFinder.java

```diff
@@ -0,0 +1,259 @@
+            // FIXME Solve the case when an artifact is not available in PNC
+            checksum.getType();
+        }
+
+        // FIXME Review logic of choosing the best artifact
+        return getBestPncArtifact(artifacts);
+    }
+
+    private Collection<Artifact> lookupPncArtifactsByChecksum(Checksum checksum) throws RemoteResourceException {
+        switch (checksum.getType()) {
+            case md5:
+                return pncClient.getArtifactsByMd5(checksum.getValue()).getAll();
+
+            case sha1:
+                return pncClient.getArtifactsBySha1(checksum.getValue()).getAll();
+
+            case sha256:
+                return pncClient.getArtifactsBySha256(checksum.getValue()).getAll();
+
+            default:
+                throw new IllegalArgumentException(
+                        "Unsupported checksum type requested! " + "Checksum type " + checksum.getType()
+                                + " is not supported.");
+        }
+    }
+
+    private int getArtifactQuality(Object obj) {
+        Artifact a = (Artifact) obj;
+        ArtifactQuality quality = a.getArtifactQuality();
+
+        switch (quality) {
+            case NEW:
+                return 1;
+            case VERIFIED:
+                return 2;
+            case TESTED:
+                return 3;
+            case DEPRECATED:
+                return -1;
+            case BLACKLISTED:
+                return -2;
+            case TEMPORARY:
+                return -3;
+            case DELETED:
+                return -4;
+            default:
+                LOGGER.warn("Unsupported ArtifactQuality! Got: " + quality);
+                return 0;
+        }
+    }
+
+    private Artifact getBestPncArtifact(Collection<Artifact> artifacts) {
+        if (artifacts == null || artifacts.isEmpty()) {
+            throw new IllegalArgumentException("No artifacts provided!");
+        }
+
+        if (artifacts.size() == 1) {
+            return artifacts.iterator().next();
+        } else {
+            return artifacts.stream()
+                    .sorted(Comparator.comparing(this::getArtifactQuality).reversed())
+                    .filter(artifact -> artifact.getBuild() != null)
+                    .findFirst()
+                    .orElse(artifacts.iterator().next());
+        }
+    }
+
+    private KojiBuild convertPncBuildToKojiBuild(PncBuild pncBuild) {
+        KojiBuild kojibuild = PncUtils.pncBuildToKojiBuild(pncBuild);
+
+        for (EnhancedArtifact artifact : pncBuild.getArtifacts()) {
+
+            KojiArchiveInfo kojiArchive = PncUtils.artifactToKojiArchiveInfo(pncBuild, artifact.getArtifact());
+            PncUtils.fixNullVersion(kojibuild, kojiArchive);
+            buildFinderUtils.addArchiveToBuild(kojibuild, kojiArchive, artifact.getFilenames());
+
+            // TODO review logic of buildZero - is it needed?
+            // KojiBuild buildZero = builds.get(new BuildSystemInteger(0, BuildSystem.none));
```

Comment:
The `buildZero` variable (build in build map with id 0) holds the archives that are not found. This is because all it returned was map id -> build and so the not found archives can be returned within the same map data structure.

Reply:
Ah, cool. That solves one open question I had. I'll do it the same way.